### PR TITLE
expose custom image interpolation kernels

### PR DESCRIPTION
**Pull request by brisvag**
# Description
Expose vispy's custom interpolation kernels at the layer level. I don't think there's a nice way to expose it in the gui, so for now it's just programmatic.

Here's the added example, using a simple magicgui slider to generate a gaussian kernel with custom size and sigma:

https://user-images.githubusercontent.com/23482191/191957080-c675e4ca-e497-4713-b7da-e74da8431df7.mp4

This PR is not yet working completely as intended, I found some issue with negative values not working (so, no sharpening/edge detection). Fix is already ready to go in at vispy/vispy#2382.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

**Comment by Czaki** [comment]
I love it, and maybe something like range modal? Some button or right click?

**Comment by brisvag** [comment]
I'm not following, can you elaborate? :)

**Comment by Czaki** [comment]
I think that interface you proposed is quite nice, but I could understand that adding the next three positions to the image controls panel may not be an optimal solution, so I suggest that it could be hidden in the collapsible part or in a modal dialog.

**Comment by brisvag** [comment]
I'm not sure we should expose it like this... A widget plugin (even builtin) would be a better approach; this way we still allow the user to provide any kernel they want, rather than providing a set of kernel ourselves.

**Comment by Czaki** [comment]
Plugin entry for kernel definition?

**Comment by brisvag** [comment]
Also...

**Comment by brisvag** [comment]
This now works as inteded for all kernels:

https://user-images.githubusercontent.com/23482191/207653777-23666a9a-40c0-4471-b0e4-f04e86e69e2f.mp4

I think we should keep this PR simple and defer GUI interface and/or plugin interfaces to a later PR.

**Comment by brisvag** [comment]
Real use case on raw, noisy cryo-EM data:

https://user-images.githubusercontent.com/23482191/207656967-a80aba1b-2e6a-45cc-b0aa-10c08daa57eb.mp4

File: napari/_vispy/layers/image.py
Comment (Czaki) [inline napari/_vispy/layers/image.py]: From where it is called when change ndisplay from 3d to 2d? Is reset function called in such situation?

**Review by alisterburt — COMMENTED** [review]
this is cool! haven't looked carefully yet but would an API like
`image.custom_2d_interpolation_kernel` not be more flexible for when we (maybe never) have custom 3D too? Otherwise we wouldn't be able to have custom 2D and 3D on the layer at the same time which feels like a reasonable expectation

**Comment by alisterburt** [comment]
q: is this interpolation or is it a filter applied prior to interpolation?

**Comment by alisterburt** [comment]
q: is this interpolation or is it a filter applied prior to interpolation?

**Comment by brisvag** [comment]
Good point, it's probably better to distinguish in 2D from 3D.

This is replacing interpolation; that's why you have to choose `custom` for the `interpolation`. Basically, instead of sampling the texture using the autogenerated kernels for the hardcoded interpolation modes, it will sample the texture based on the given kernel by convolving.

**Comment by brisvag** [comment]
what about `custom_kernel_interpolation2d`, closer to the `interpolation2d` parameter?

File: napari/_vispy/layers/image.py
Comment (brisvag) [reply napari/_vispy/layers/image.py]: yes, at the end of `_on_display_change`, reset is called.

**Comment by alisterburt** [comment]
> This is replacing interpolation; that's why you have to choose custom for the interpolation. Basically, instead of sampling the texture using the autogenerated kernels for the hardcoded interpolation modes, it will sample the texture based on the given kernel by convolving.

Sorry, I don't understand in detail how the interpolation works here and must be missing something - if I sample a position in continuous coordinates `(x, y)` in an image how exactly is the kernel used to yield the value for that position? 

To me, the kernel defines weights of different image pixels in a local region - I don't see how it depends on the position of the continuous coordinate in a way that changes the output screen pixel value, what am I missing?

**Comment by brisvag** [comment]
Well, yes, you're right, they are not really equivalent; what is really happening is that the *sampling* is set to the "harware" linear, and on top of that we do this kernel interpolation.

The reason why I went this way is that it was too computationally expensive to do both texture interpolation (which requires issuing a texture sample call for each texel needed by the interpolation kernel) and this extra kernel pass (which would multiply by `kernel.size` the number of sampling calls). It also required some extra refactoring, so I opted for the simpler implementation.

**Comment by alisterburt** [comment]
> Well, yes, you're right, they are not really equivalent; what is really happening is that the _sampling_ is set to the "harware" linear, and on top of that we do this kernel interpolation.

and this convolution with the kernel is applied in screen pixel space, after the linear sampling?

**Comment by brisvag** [comment]
The more I answer your questions, the more I realize this is a bit weird xD

> and this convolution with the kernel is applied in screen pixel space, after the linear sampling?

Not quite: it happens after the linear sampling, but the kernel is applied in image pixels, not screen pixels. Zooming the image does not change the looks.
I's easier to explain with an image:

image

The big square is the image, and the grid represents texels. However, we have more fragments than texels because we're zoomed in. If we pick a random fragment (the point in the center of the smaller square), the resulting color is given by sampling each point on the the (3x3 in this case) kernel and doing the usual weighted sum. This means that a slightly offset fragment will use slightly different "raw samples" (due to linear interpolation).

**Comment by alisterburt** [comment]
gotcha, thanks for the explanation, that's (very weird, but...) cool! 

Would be cool to do something like this in 3D during the refine loop of the mip/minip shaders to get 'efficient' downsampling on the fly

**Comment by alisterburt** [comment]
It would be really cool to get this in! Before merge we should have:
- the differentiation between 2D and 3D
- an explanatory doc explaining how to use the feature

I would prefer `custom_interpolation_kernel_2d` over `custom_kernel_interpolation_2d`

**Comment by brisvag** [comment]
updated the name and added a bit of docstring to the `interpolation` argument. Do you think we need more?

File: examples/image_custom_kernel.py
Comment (alisterburt) [inline examples/image_custom_kernel.py]: explanation here

File: napari/layers/image/image.py
Comment (alisterburt) [inline napari/layers/image/image.py]: almost! not quite clear enough I think, how about...

'custom' is a special mode for 2D interpolation in which a regular grid of samples are taken from the texture around a position using 'linear' interpolation before being multiplied with a custom interpolation kernel (provided with 'custom_interpolation_kernel_2d').

Something else that's a little unclear to me is where the center of the kernel is if the kernel has even sidelength, do. you know what happens there? :)

**Review by alisterburt — APPROVED** [review]
Given how unintuitive it is, I'd like it to be really obvious what's actually happening under the hood when you set this, have suggested a change - for me this is good to go once the explanation is in place, nice work!

File: napari/layers/image/image.py
Comment (brisvag) [reply napari/layers/image/image.py]: The center of the kernel is always centered, so the exact center is not sampled.

**Comment by brisvag** [comment]
Updated the docs!

**Review by alisterburt — APPROVED** [review]
looks good, thanks for the extra explanation!

**Review by Czaki — COMMENTED** [review]
Maybe some test?

**Comment by brisvag** [comment]
What test do you think would help here? Just to check if the values are fowrarded correctly to vispy?

**Comment by Czaki** [comment]
> What test do you think would help here? Just to check if the values are fowrarded correctly to vispy?

If it will be possible, then great. But at least there should be API call that checks if it does not crash.

**Pull request merged**
